main.tf

```diff
@@ -118,26 +188,73 @@ resource "ibm_is_instance" "vsi" {
     ]
   }
 
-  primary_network_interface {
-    subnet = each.value.subnet_id
-    security_groups = flatten([
-      (var.create_security_group ? [ibm_is_security_group.security_group[var.security_group.name].id] : []),
-      var.security_group_ids,
-      (var.create_security_group == false && length(var.security_group_ids) == 0 ? [data.ibm_is_vpc.vpc.default_security_group] : []),
-    ])
-    allow_ip_spoofing = var.allow_ip_spoofing
-    dynamic "primary_ip" {
-      for_each = var.manage_reserved_ips ? [1] : []
-      content {
-        reserved_ip = ibm_is_subnet_reserved_ip.vsi_ip[each.value.name].reserved_ip
+  # Primary Virtual Network Interface
+  dynamic "primary_network_attachment" {
+    for_each = var.use_legacy_network_interface ? [] : [1]
+    content {
+      name = "${each.value.subnet_name}-eth0"
+      virtual_network_interface {
+        id = ibm_is_virtual_network_interface.primary_vni[each.key].id
+      }
+    }
+  }
+
+  # Additional Virtual Network Interface
+  dynamic "network_attachments" {
+    for_each = {
+      for k in var.secondary_subnets : k.zone => k
+      if k.zone == each.value.zone && !var.use_legacy_network_interface
+    }
+    content {
+      name = network_attachments.value.name
+      virtual_network_interface {
```

Comment:
I think all of the VNIs should be managed by terraform outside the VSI block, just like you have done with the primary. That will make it much easier to leave them alone and simply reattach if the VSI gets recreated, and they will retain all of their settings and IPs. Basically we want the VSI block here to manage as little as possible and we would "attach" resources managed outside the block instead. This way if a rollback occurs (or the VSI gets recreated for any reason) it will not destroy some of these other attachments (like VNIs) so that they maintain their state when reattached.